swarm/src/handler.rs

```diff
@@ -334,98 +331,69 @@ pub enum ProtocolsChange<'a> {
 }
 
 impl<'a> ProtocolsChange<'a> {
-    /// Compute the [`ProtocolsChange`] that results from adding `to_add` to `existing_protocols`.
-    ///
-    /// Returns `None` if the change is a no-op, i.e. `to_add` is a subset of `existing_protocols`.
-    pub(crate) fn add(
-        existing_protocols: &'a HashSet<StreamProtocol>,
-        to_add: &'a HashSet<StreamProtocol>,
-    ) -> Option<Self> {
-        let mut actually_added_protocols = to_add.difference(existing_protocols).peekable();
-
-        actually_added_protocols.peek()?;
-
-        Some(ProtocolsChange::Added(ProtocolsAdded {
-            protocols: actually_added_protocols,
-        }))
-    }
-
-    /// Compute the [`ProtocolsChange`] that results from removing `to_remove` from `existing_protocols`.
-    ///
-    /// Returns `None` if the change is a no-op, i.e. none of the protocols in `to_remove` are in `existing_protocols`.
-    pub(crate) fn remove(
-        existing_protocols: &'a HashSet<StreamProtocol>,
-        to_remove: &'a HashSet<StreamProtocol>,
-    ) -> Option<Self> {
-        let mut actually_removed_protocols = existing_protocols.intersection(to_remove).peekable();
-
-        actually_removed_protocols.peek()?;
-
-        Some(ProtocolsChange::Removed(ProtocolsRemoved {
-            protocols: Either::Right(actually_removed_protocols),
-        }))
-    }
-
     /// Compute the [`ProtocolsChange`]s required to go from `existing_protocols` to `new_protocols`.
-    pub(crate) fn from_full_sets(
-        existing_protocols: &'a HashSet<StreamProtocol>,
-        new_protocols: &'a HashSet<StreamProtocol>,
-    ) -> SmallVec<[Self; 2]> {
-        if existing_protocols == new_protocols {
-            return SmallVec::new();
-        }
-
-        let mut changes = SmallVec::new();
-
-        let mut added_protocols = new_protocols.difference(existing_protocols).peekable();
-        let mut removed_protocols = existing_protocols.difference(new_protocols).peekable();
-
-        if added_protocols.peek().is_some() {
-            changes.push(ProtocolsChange::Added(ProtocolsAdded {
-                protocols: added_protocols,
-            }));
+    pub(crate) fn from_full_sets<T: AsRef<str>>(
+        existing_protocols: &mut HashMap<AsStrHashEq<T>, bool>,
```

Comment:
Can you explain what is happening with this `HashMap` here? The reason we currently use `.difference` of `HashSet` is because it is completely lazy. It does nothing until somebody iterates over these events. Many protocols don't look at these events so this should be fairly efficient.

I like your design of the `AsStrHashEq` wrapper. Can we use that in combination with the `HashSet`-based design? Perhaps we can even lazily parse the `&str` into a `StreamProtocol` as part of the iterator. That would avoid even more allocations in the common case.

Reply:
> The reason we currently use .difference of HashSet is because it is completely lazy.

Moving pointers to a vector is not a bottleneck, populating the `HashSet` to make the diff is.

> Perhaps we can even lazily parse the &str into a StreamProtocol as part of the iterator. That would avoid even more allocations in the common case.

Not unless we introduce generic parameter to `ProtocolsAdded`.

Reply:
Reaoning behind the code is to perform a diff with least amount of hashmap operations since profiler pointed me to hashmap taking most of the time

Reply:
> Reaoning behind the code is to perform a diff with least amount of hashmap operations since profiler pointed me to hashmap taking most of the time

Can you document how it works please?

I understand that it is a performance optimisation (the entire PR is) but I don't understand how it works and I refuse to reverse-engineer it from reading the code :)

Some questions that I'd like to be answered:

- What does the `boolean` do?
- Why do we need a `HashMap`?
- What are we counting?
- Why is `count` an `Option`?
- What is a "valid_value" and what is is valid for? Can it be invalid?
- If all values in the map are the same, why do we even need to store them in the `HashMap`?

Reply:
> > The reason we currently use .difference of HashSet is because it is completely lazy.
> 
> Moving pointers to a vector is not a bottleneck, populating the `HashSet` to make the diff is.

Interesting! I remember from my own benchmarking in this area that one of the bottlenecks was also that we would _always_ construct a `StreamProtocol` using `try_from_owned` which I believe we are avoiding with the design of the wrapper.

I'd very interested in seeing benchmark numbers that compare:

- What is currently on master
- A `HashSet`-based design using `AsStrHashEq`
- This PR

My intuition is that populating a new `HashSet` from all protocols is negligible compared to allocating for each `StreamProtocol`. I suspect that the actual performance improvement here is combing from using `AsStrHashEq`. If that is the case, I'd strongly favor the `HashSet`-based design as it is much easier to understand.

We could even get rid of those allocations by collecting the protocols into an existing `HashSet`. Then the cost should come down to a few comparison and hash operations which should be extremely cheap.

Reply:
> A `HashSet`-based design using `AsStrHashEq`

I just tried this and couldn't quite get it to work unfortunately. Without adding a type parameter to `ProtocolsAdded`, we cannot directly reference a `HashSet<AsStrHashEq<T>>` :(

Reply:
Clearing and reinserting things to a hashset is actually expensive here, even if I reuse allocations.

> I just tried this and couldn't quite get it to work unfortunately. Without adding a type parameter to ProtocolsAdded, we cannot directly reference a HashSet<AsStrHashEq<T>> :(

Yes, I also wanted to do that, only possibility I see is using `&mut dyn` somehow, though that will be super unreadable.

Reply:
If you are curious, benchmark with hashset is mentioned in https://github.com/libp2p/rust-libp2p/pull/5026#issuecomment-1868828607, it is faster but still twice as slow then this

Reply:
I cleaned the code up, it was really confusing, hope this makes it more clear